flowauth/frontend/src/util/util.js

```diff
@@ -26,80 +27,144 @@ function zip() {
-      split.forEach((k, ix) => {
-        if (!(k in obj)) {
-          obj[k] = {};
-        }
-        obj = obj[k];
-        obj["parent"] = parents.join("&");
-      });
-      obj["full_path"] = sub_scope;
-      last_split = split;
-      parents.push(sub_scope);
-    });
-  });
-  return nested;
+
+
+
+
+  /** Takes an array of scopes, most of which are triplets in the form admin_level:top_level_query:dependent_query
+  * and converts them into a graph keyed on admin_level -> top_level_query -> dependent_query
+  */
+export function scopesGraph(scopes_obj) {
+  var scopes_array
+  if (!Array.isArray(scopes_obj)){
+    scopes_array = Object.keys(scopes_obj)
+  }
+  else{
+    scopes_array = scopes_obj
+  }
+  const triplet_scopes = scopes_array.filter(row => row.includes(":"))  // It's this or a regex
+  const split_scopes = triplet_scopes.map(row => row.split(":"))
+  const scopes_tree = {}
+  for (const row of split_scopes){
+    var [ admin_level, tl_query, inner_query ] = row
+
+    if (!(Object.keys(scopes_tree).includes(admin_level))){
+      scopes_tree[admin_level] = {}
+    }
+    if (!(Object.keys(scopes_tree[admin_level]).includes(tl_query))){
+      scopes_tree[admin_level][tl_query] = {}
+    }
+    scopes_tree[admin_level][tl_query][inner_query] = true
+  }
+  const single_scopes=scopes_array.filter(row => !row.includes(":"))
+    .map(scope => ({[scope]:true}))
+    .reduce((p_x, x) => ({...p_x, ...x}))
+  const all_scopes = ({...scopes_tree, ...single_scopes})
+  console.log(all_scopes)
+  return all_scopes
 }
 
-export function jsonify(tree, labels, enabled, enabledKeys) {
-  const parent = tree.parent;
-
-  const list = Object.keys(tree)
-    .filter((k) => k !== "parent" && k !== "full_path")
-    .map((k) => {
-      const ll = labels.concat([k]);
-      const v = tree[k];
-      const val = parent ? [parent, ll.join(".")].join("&") : ll.join(".");
-      if (
-        Object.keys(v).filter((k) => k !== "parent" && k !== "full_path")
-          .length === 0
-      ) {
-        const value = [parent, v.full_path].join("&");
-        return {
-          label: k,
-          value: value,
-          parents: parent,
-          enabled: enabled.includes(value),
-        };
-      } else {
-        const children = jsonify(
-          v,
-          v.parent === parent ? ll : [],
-          enabled,
-          enabledKeys
-        );
-        const allEnabled = children.every((child) => child.enabled);
-        if (!allEnabled) {
-          children
-            .filter((child) => child.enabled)
-            .forEach((child) => enabledKeys.push(child.value));
-        }
-        return {
-          label: k,
-          value: val,
-          children: children,
-          parents: parent,
-          enabled: allEnabled,
-        };
+
+function jsonify_inner(tree, label, value, enabled){
+  const things = Object.keys(tree).map((branch, index) => {
+    const this_label = label ==="" ? branch: [label, branch].join(".")
```

Comment:
It's a minor point, but I'm not sure we need the label to be the full tree path at each level of the cascade. I.e. instead of
```
[] agg_unit -> [] agg_unit.tl_query -> [] agg_unit.tl_query.sub_query
```
I think it would be sufficient (and more concise) to display
```
[] agg_unit -> [] tl_query -> [] sub_query
```
(the layout of the MultiCascader itself shows the full path, so I don't think the labels need to).